static/analysis.go

```diff
@@ -1,32 +1,43 @@
 package static
 
 import (
+	"fmt"
 	"go/constant"
 	"go/types"
 
 	"github.com/dogmatiq/configkit"
 	"github.com/dogmatiq/configkit/internal/entity"
 	"github.com/dogmatiq/configkit/internal/typename/gotypes"
+	"github.com/dogmatiq/configkit/message"
 	"golang.org/x/tools/go/ssa"
 )
 
 // analyzeApplication analyzes a type that implements the dogma.Application
-// interface to deduce it's application configuration.
-func analyzeApplication(pkg *ssa.Package, typ types.Type) configkit.Application {
+// interface to deduce its application configuration.
+func analyzeApplication(prog *ssa.Program, typ types.Type) configkit.Application {
 	app := &entity.Application{
-		TypeNameValue: gotypes.NameOf(typ),
+		TypeNameValue:     gotypes.NameOf(typ),
+		HandlersValue:     configkit.HandlerSet{},
+		MessageNamesValue: configkit.EntityMessageNames{},
 	}
 
-	fn := pkg.Prog.LookupMethod(
-		typ,
-		pkg.Pkg,
-		"Configure",
-	)
+	pkg := getTypePkg(typ)
+	fn := prog.LookupMethod(typ, pkg, "Configure")
 
 	for _, c := range findConfigurerCalls(fn) {
+		args := c.Common().Args
+
 		switch c.Common().Method.Name() {
 		case "Identity":
 			app.IdentityValue = analyzeIdentityCall(c)
+		case "RegisterAggregate":
+			app.HandlersValue.Add(
+				analyzeHandler(
+					prog,
+					args[0].(*ssa.MakeInterface).X.Type(),
```

Comment:
Can you always rely on the argument being one of these `MakeInterface` types? What if, for example, the argument was a variable of type `dogma.AggregateMessageHandler` ?

Reply:
Or `nil`, for that matter. Obviously this isn't normal but it would still compile and we'd want to make sure that this package didn't choke on that input.

Reply:
Good point, I will write the code and tests to cover for this scenarios.

Reply:
Ok, even if we assign an instance of aggregate to a variable of type `dogma.AggregateMessageHandler` and pass it as an argument to the method, `ssa` packages still interprets that as `*ssa.MakeInterface` instruction.

However, passing `nil` is treated as `*ssa.Const` which causes the code above to explode. 

Should we do a check similar to this and write tests to address this scenario?

```go
            case "RegisterAggregate":
			if mi, ok := args[0].(*ssa.MakeInterface); ok {
				app.HandlersValue.Add(
					analyzeHandler(
						prog,
						mi.X.Type(),
						configkit.AggregateHandlerType,
					),
				)
			}
		}
```

Reply:
Can you try what happens if there is a typed `const` that is a real implementation of `AggregateMessageHandler`?

like ...

```go
type SomeHandler string

const Handler SomeHandler = "..."

func (Blah) Configure(........
```

Because I wonder if that is a `*ssa.Const` or not, since it obvious IS a const, but also implements the interface.

Reply:
Funny thing that the type above still ends up as `*ssa.MakeInterface`: 
```golang
&ssa.MakeInterface{
    register: ssa.register{
        anInstruction: ssa.anInstruction{
			block:(*ssa.BasicBlock)(0xc00364c790),
		},
		num:3,
		typ:(*types.Named)(0xc001200140),
		pos:0,
		referrers:[]ssa.Instruction{
			(*ssa.Call)(0xc003f92280),
		},
	},
	X:(*ssa.UnOp)(0xc005177500),
}
```

Reply:
🤷

Reply:
Ok, cool.. I guess it represents the "boxing" of the type into the underlying interface struct?

I think the `ok` style type assertion is the way to go then.

Reply:
Yep, boxing would be exactly what's happening here in my mind too. I will still throw some tests for passing `nil` in `RegisterAggregate()`, `ConsumesCommandType()`, and `ProducesEventType()` then.

Reply:
Ok, added type assertions and tests in https://github.com/dogmatiq/configkit/pull/123/commits/6c01fa7d4e0c70a83b8c062dfd27e415c23177fc.

Reply:
:+1: